tests/test_pagination.py

```diff
@@ -152,16 +149,66 @@ def pagination_test(self):
 
                     # Verify we have more fields sent to the target than just automatic fields (this is set above)
-                    self.assertEqual(auto_fields.difference(actual_keys),
-                                     set(), msg="A paginated synced stream has a record that is missing expected fields.")
+                    self.assertEqual(set(), auto_fields.difference(actual_keys),
+                                     msg="A paginated synced stream has a record that is missing expected fields.")
 
-                # TODO ADD CHECK ON IDS
                 # Verify by pks that the data replicated matches what we expect
+                primary_keys = self.expected_primary_keys().get(stream)
+                pk = list(primary_keys)[0] if primary_keys else None
+
+                if pk:
+                    # Verify that actual records were in our expectations
+                    for actual_record in actual_records:
+                        stream_expected_records = [record.get(pk) for record in expected_records.get(stream)
+                                                   if actual_record.get(pk) == record.get(pk)]
+                        self.assertTrue(len(stream_expected_records),
+                                        msg="An actual record is missing from our expectations: \nRECORD: {}".format(actual_record))
+                        self.assertEqual(1, len(stream_expected_records),
+                                         msg="A duplicate record was found in our expectations for {}.".format(stream))
+                        stream_expected_record = stream_expected_records[0]
+                        self.assertEqual(stream_expected_record, actual_record.get(pk))
+
+                    # Verify that our expected records were replicated by the tap
+                    for expected_record in expected_records.get(stream):
+                        stream_actual_records = [record.get(pk) for record in actual_records
+                                                 if expected_record.get(pk) == record.get(pk)]
+                        self.assertTrue(len(stream_actual_records),
+                                        msg="An expected record is missing from the sync: \nRECORD: {}".format(expected_record))
+                        self.assertEqual(1, len(stream_actual_records),
+                                         msg="A duplicate record was found in the sync for {}.".format(stream))
+                        stream_actual_record = stream_actual_records[0]
+                        self.assertEqual(expected_record.get(pk), stream_actual_record)
+
+                elif stream == 'inventories': # We can not compare by pks for this stream so we make a less strict assertion
+
+                    # Verify the expected number of records were replicated
+                    self.assertEqual(len(expected_records.get(stream)), len(actual_records))
```

Comment:
Couldn't we assert this in all cases and pull it out of the if block?

Reply:
Yes we could, but the assertions above will catch if the lengths are different and have the benefit of printing the extra/missing record. If we added this prior to the assertions above it would prevent us from logging the diff. If we added it after it would never fail because we would fail on line 180 or 169 first.

Reply:
Well I suppose it doesn't hurt the test to pull it from the elif and change that to an if. I can do that